### PR TITLE
bpo-42808: Add PyType_Type.tp_vectorcall for type(obj) performance

### PR DESCRIPTION
**Pull request by sweeneyde**
https://bugs.python.org/issue42808

### CODE REVIEW

File: Objects/typeobject.c
Comment (corona10) [inline Objects/typeobject.c]: nit: we can use `Py_NewRef()` in this case?

**Comment by sweeneyde** [comment]
@markshannon As code owner, do you think there is any worth to this?

**Comment by markshannon** [comment]
Yes, I do. `type(x)` is quite common in Python code (unlike the 3 argument form).

**Comment by markshannon** [comment]
LGTM. Let's rerun the tests before merging as they are a month old.

**Comment by sweeneyde** [comment]
Any update on this? I'm seeing some nice benchmarks:

```
.\python.bat -m pyperf timeit -s "import collections, itertools" "collections.deque(map(type, itertools.repeat(17, 1_000_000)), maxlen=0)"
Before: Mean +- std dev: 26.0 ms +- 0.8 ms
After:  Mean +- std dev: 11.6 ms +- 0.1 ms

.\python.bat -m pyperf timeit "type(17)"
Before: 51.6 ns +- 3.5 ns
After:  33.8 ns +- 2.1 ns
```

**Comment by sweeneyde** [comment]
@markshannon Any update on this?

**Review by corona10 — APPROVED** [review]
LGTM I will merge this PR after tests are passed ;)
cc @markshannon

**Comment by corona10** [comment]
Buildbot failure is unrelated to this PR

**Pull request merged**
